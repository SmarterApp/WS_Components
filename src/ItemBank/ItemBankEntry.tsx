import * as React from "react";
import {
  Accordion,
  ToolTip,
  ItemRevisionModel,
  SectionModel,
  ItemEntryTable
} from "../index";
import { CsvEntry } from "../CsvEntry/CsvEntry";

export interface ItemBankEntryProps {
  updateItems: (items: ItemRevisionModel[]) => void;
  sections: SectionModel[];
  items: ItemRevisionModel[];
}

export interface ItemBankEntryState {
  csvIsOpen: boolean;
  itemsEntryIsOpen: boolean;
}

export class ItemBankEntry extends React.Component<
  ItemBankEntryProps,
  ItemBankEntryState
> {
  constructor(props: ItemBankEntryProps) {
    super(props);
    this.state = {
      csvIsOpen: false,
      itemsEntryIsOpen: false
    };
  }

  onCsvClick = () => {
    this.setState({
      csvIsOpen: true
    });
  };

  onCsvBlur = () => {
    this.setState({
      csvIsOpen: false,
      itemsEntryIsOpen: true
    });
  };

  renderCsvEntry() {
    return (
      <Accordion accordionTitle="CSV Entry" isOpen={this.state.csvIsOpen}>
        <div className="csv-entry-container section">
          <div className="csv-entry-wrapper">
            <CsvEntry
              onItemsUpdate={this.props.updateItems}
              onBlur={this.onCsvBlur}
              onClick={this.onCsvClick}
            />
          </div>
        </div>
      </Accordion>
    );
  }

  renderTableEntry() {
    const { items, updateItems, sections } = this.props;

    return (
      <Accordion
        accordionTitle="Items Entry"
        isOpen={this.state.itemsEntryIsOpen}
      >
        <ItemEntryTable
          itemRows={items}
          onItemsUpdate={updateItems}
          sections={sections}
        />
      </Accordion>
    );
  }

  render() {
    return (
      <div>
        {this.renderCsvEntry()}
        {this.renderTableEntry()}
      </div>
    );
  }
}
