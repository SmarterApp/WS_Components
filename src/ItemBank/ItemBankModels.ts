import {
  ItemModel,
  AboutItemRevisionModel,
  GradeLevel,
  RevisionModel
} from "@src/index";

export interface ItemRevisionModel {
  itemKey?: number;
  bankKey?: number;
  section?: string;
  revision?: string;
  isaap?: string;
}

export function getItemBankName(itemRevisionModel: ItemRevisionModel) {
  let value = "";
  if (itemRevisionModel.bankKey && itemRevisionModel.itemKey) {
    value = `${itemRevisionModel.bankKey}-${itemRevisionModel.itemKey}`;
  }

  return value;
}

export function itemRevisionKey(itemRevisionModel: ItemRevisionModel) {
  return `${itemRevisionModel.bankKey}-${itemRevisionModel.itemKey}-${
    itemRevisionModel.section
  }`;
}

export function validItemRevisionModel(itemRevisionModel: ItemRevisionModel) {
  let value = false;
  if (
    itemRevisionModel.itemKey &&
    itemRevisionModel.bankKey &&
    itemRevisionModel.section
  ) {
    value = true;
  }

  return value;
}

export function getItemBankIndex(
  currentItem: ItemRevisionModel,
  items: ItemRevisionModel[]
) {
  const key = itemRevisionKey(currentItem);

  return items.findIndex(it => itemRevisionKey(it) === key);
}

export function getNextItemBank(
  currentItem: ItemRevisionModel,
  items: ItemRevisionModel[]
): ItemRevisionModel | undefined {
  const currentIdx = getItemBankIndex(currentItem, items);
  let nextItem: ItemRevisionModel | undefined;

  const nextIdx = currentIdx + 1;
  if (
    nextIdx >= 0 &&
    nextIdx < items.length &&
    validItemRevisionModel(items[nextIdx])
  ) {
    nextItem = items[nextIdx];
  }

  return nextItem;
}

export function getPreviousItemBank(
  currentItem: ItemRevisionModel,
  items: ItemRevisionModel[]
): ItemRevisionModel | undefined {
  let previousItem: ItemRevisionModel | undefined;
  const currentIdx = getItemBankIndex(currentItem, items);

  const prevIdx = currentIdx - 1;
  if (prevIdx >= 0 && validItemRevisionModel(items[prevIdx])) {
    previousItem = items[prevIdx];
  }

  return previousItem;
}

export interface SectionModel {
  key: string;
  value: string;
}

export interface AccessibilityRevisionModel {
  gradeLevel: GradeLevel;
  subject: string;
  interactionType: string;
  allowCalculator?: boolean;
  isPerformance?: boolean;
}
