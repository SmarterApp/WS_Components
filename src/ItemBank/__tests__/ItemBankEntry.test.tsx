import * as React from "react";
import { shallow } from "enzyme";
import { ItemBankEntry } from "../ItemBankEntry";
import { SectionModel, ItemRevisionModel } from "../ItemBankModels";

const section: SectionModel = {
  key: "187-3000",
  value: "math"
};

const item: ItemRevisionModel = {
  itemKey: 187,
  bankKey: 3000,
  section: "math",
  revision: "asdge",
  isaap: ""
};

const items: ItemRevisionModel[] = [item];

const sections: SectionModel[] = [section];

describe("ItemBankEntry", () => {
  const wrapper = shallow(
    <ItemBankEntry
      updateItems={items => jest.fn()}
      sections={sections}
      items={items}
    />
  );

  it("matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
