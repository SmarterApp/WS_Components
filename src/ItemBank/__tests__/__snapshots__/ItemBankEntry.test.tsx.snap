// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemBankEntry matches snapshot 1`] = `
<div>
  <Accordion
    accordionTitle="CSV Entry"
    contentItem={
      <div
        className="csv-entry-container section"
      >
        <div
          className="csv-entry-wrapper"
        >
          <CsvEntry
            onItemsUpdate={[Function]}
          />
        </div>
      </div>
    }
  />
  <Accordion
    accordionTitle="Items Entry"
    contentItem={
      <ItemEntryTable
        itemRows={
          Array [
            Object {
              "bankKey": 3000,
              "isaap": "",
              "itemKey": 187,
              "revision": "asdge",
              "section": "math",
            },
          ]
        }
        onItemsUpdate={[Function]}
        sections={
          Array [
            Object {
              "key": "187-3000",
              "value": "math",
            },
          ]
        }
      />
    }
  />
</div>
`;

exports[`ItemBankViewer matches snapshot 1`] = `
<div>
  <Accordion
    accordionTitle="CSV Entry"
    contentItem={
      <div
        className="csv-entry-container section"
      >
        <div
          className="csv-entry-wrapper"
        >
          <CsvEntry
            onItemsUpdate={[Function]}
          />
        </div>
      </div>
    }
  />
  <Accordion
    accordionTitle="Items Entry"
    contentItem={
      <ItemEntryTable
        itemRows={
          Array [
            Object {
              "bankKey": 3000,
              "isaap": "",
              "itemKey": 187,
              "revision": "asdge",
              "section": "math",
            },
          ]
        }
        onItemsUpdate={[Function]}
        sections={
          Array [
            Object {
              "key": "187-3000",
              "value": "math",
            },
          ]
        }
      />
    }
  />
</div>
`;
